Remove global symbols: Avoid using the global namespace if possible

practicalswift opened the pull request:
Remove global symbols: Avoid using the global namespace if possible.

Partially resolves #15612 ("Reduce the number of global symbols used").

Change in global symbols as reported by `nm bitcoind` before vs after:

```
$ diff -u <(nm src/bitcoind-before | c++filt | grep -E '^[0-9a-f]+ [A-Z] ' | cut -f3- -d' ' | sort -u) \
          <(nm src/bitcoind-after  | c++filt | grep -E '^[0-9a-f]+ [A-Z] ' | cut -f3- -d' ' | sort -u) \
          | grep -E '^[+-][^+-]'
-boundSockets
-cs_warnings
-eventHTTP
-fFeeEstimatesInitialized
-fLargeWorkForkFound
-fLargeWorkInvalidChainFound
-pathHandlers
-strMiscWarning[abi:cxx11]
-threadHTTP
```

MarcoFalke commented on src/httpserver.cpp:
```diff
@@ -137,16 +137,20 @@ struct HTTPPathHandler
+    std::vector<HTTPPathHandler> pathHandlers;
+    //! Bound listening sockets
+    std::vector<evhttp_bound_socket *> boundSockets;
+}
```
Why can't you wrap everything above into a namespace? Also should not change the indentation. I'd suggest to use clang-format

practicalswift replied:
Good point! Fixed!

With this change `HTTPPathHandler`, `WorkQueue` and `HTTPWorkItem` are no longer global.

Please re-review!

promag commented:
Concept ACK, change looks good.

MarcoFalke commented:
utACK b112097f8e5f96e96db47e2eef9870d6f0431677. Also checked that the binary is 9kB smaller.

Empact commented:
utACK https://github.com/bitcoin/bitcoin/pull/15622/commits/b112097f8e5f96e96db47e2eef9870d6f0431677

promag commented on src/init.cpp:
```diff
@@ -71,13 +71,15 @@
 
+namespace {
 bool fFeeEstimatesInitialized = false;
 static const bool DEFAULT_PROXYRANDOMIZE = true;
```
I think you could drop `static`?

practicalswift replied:
Yes, absolutely but I didn't want to change any existing lines to make review super trivial. Let me know if that decision is blocking an utACK from you and I'll consider reconsidering :-)

Empact replied:
With or without the declaration seems they'll have internal linkage (via unnamed namespace) and static storage duration (via namespace scope).

Mildly in favor of removal on the grounds that a distinction without a difference in this case can be misleading.

> `static` When used in a declaration of an object, it specifies static storage duration. When used in a declaration at namespace scope, it specifies internal linkage.
> static storage duration. All objects declared at namespace scope (including global namespace) have this storage duration, plus those declared with static or extern.
https://en.cppreference.com/w/cpp/language/storage_duration

> Unnamed namespace definition. Its members have potential scope from their point of declaration to the end of the translation unit, and have internal linkage.
https://en.cppreference.com/w/cpp/language/namespace#Unnamed_namespaces

practicalswift commented:
Now removing redundant `static`:s in unnamed namespaces as requested by @promag and @Empact.

Please re-review :-)

promag commented:
utACK 9638da7.

lucayepa commented:
I think we can add some language on this policy in doc/developer-notes.md. It will help new developers. I agree on a zero-globals policy, enforced with singletons, or with a "state" class that is a container of all the globals, or with a dedicated namespace. This way, every time a new global has to be added (rarely), we can see a clear change in a single place. In my mind this policy should not be enforced from now, but should be accepted, then written in doc/developer-notes.md or elsewhere, and enforced after a couple of years, when all the open PRs are already compliant.

I've just described this method of long grace periods in my [comment on #15465](https://github.com/bitcoin/bitcoin/issues/15465#issuecomment-475914641).

practicalswift commented:
@lucayepa I would like to restrict the scope of this PR to these specific code changes which have already been (ut)ACK:ed.

Empact commented:
utACK https://github.com/bitcoin/bitcoin/pull/15622/commits/9638da76386ba1e6f610f26a5bc9b88825087bdc

MarcoFalke commented on src/init.cpp:
```diff
@@ -71,13 +71,15 @@
 // Dump addresses to banlist.dat every 15 minutes (900s)
-static constexpr int DUMP_BANS_INTERVAL = 60 * 15;
+constexpr int DUMP_BANS_INTERVAL = 60 * 15;
+} // namespace
```
Why is this changed? What is the difference between `static const(expr) int` and `const int` in a namespace?

We don't change code for stylistic reasons.

MarcoFalke replied:
Also, I cant see a reason to not have literals in the global namespace

sipa replied:
static inside a anonymous namespace is redundant.

practicalswift replied:
Please see also https://github.com/bitcoin/bitcoin/pull/15622#discussion_r267802687

Let me know if you find any reason to revert to the original version.

practicalswift replied:
> Also, I cant see a reason to not have literals in the global namespace

If we want to share say `DUMP_BANS_INTERVAL` to another compilation unit the recommended way to do it is via `#include "init.h"`, no?

Given the amount of hate we get due to the said environmental impact of SHA256² we don't want to make things worse by polluting also the global namespace! :-)

MarcoFalke replied:
>  static inside a anonymous namespace is redundant.

ftfy: anonymous namespace is redundant when the constexpr is already static

sipa commented:
It's a bit confusing to call this"Remove globals"; a global inside a namespace is still a global.

practicalswift commented:
@sipa I've now clarified that I this refers to global **symbols** :-)

practicalswift commented:
utACK:s from @Empact and @promag, one stale utACK from @MarcoFalke and zero NACK:s

@MarcoFalke, are we ready to move forward with this one to get a slightly less polluted global namespace? :-)

ajtowns commented:
I would have thought it would be better to write these as `static ...` than `namespace { \n ... \n }` -- you can grep for the former relatively easily, but not really the latter, and it seems more consistent with the current style... It gives a +9-9 diff which seems better than +17-9...

Either way, I don't see the point of dropping `static` just because it's redundant; that seems like needless churn. After stripping the binary, i don't see any size difference in bitcoind amongst these variants on Linux with clang fwiw.

practicalswift committed:
Remove global symbols: Avoid using the global namespace if possible

Rename CCriticalSection to RecursiveMutex (both are AnnotatedMixin<std::recursive_mutex>)

```
$ git grep -E '(typedef|using).*(CCriticalSection|RecursiveMutex)'
src/sync.h:using RecursiveMutex = AnnotatedMixin<std::recursive_mutex>;
src/sync.h:typedef AnnotatedMixin<std::recursive_mutex> CCriticalSection;
```

practicalswift commented:
@ajtowns I've now reverted to the original version of the PR which is in accordance with your suggestions. Please re-review :-)

@promag @MarcoFalke @Empact Would you mind re-reviewing? :-)

MarcoFalke commented on src/warnings.cpp:
```diff
@@ -8,10 +8,10 @@
-std::string strMiscWarning GUARDED_BY(cs_warnings);
-bool fLargeWorkForkFound GUARDED_BY(cs_warnings) = false;
-bool fLargeWorkInvalidChainFound GUARDED_BY(cs_warnings) = false;
+static CCriticalSection cs_warnings;
```
style nit:

```suggestion
static RecursiveMutex cs_warnings;
```

practicalswift replied:
Done!

MarcoFalke commented:
Please squash the second commit, no need to have a separate commit for a single line stylistic fixup

practicalswift commented:
@MarcoFalke Done! Please re-review :-)
